### ISSUE TITLE
2 parses when there should be 1

### ISSUE THREAD

**Issue by alecmocatta** [Hardmath123/nearley#60]
Thanks for your lightning quick response on my last bug report Hardmath123. Another suspected bug now:

The grammar below produces 1 parse for `$a=1;` but 2 parses for `$a =1;`

Unless I'm overlooking something in my grammar, I can't see why this would be the case.

```
program -> _ block {% function(d) { return d[1]; } %}
block -> (statement _):* {% function(d) { return ["block", d[0].map(function(s){return s[0];})]; } %}
statement -> expression _ ";" {% id %}
expression -> expression _ ("="|"=="|"!="|">"|"<"|"<="|">=") _ sum {% function(d) { return ["operation", d[0], d[2], d[4]]; } %} | sum {% id %}
sum -> sum ("*"|"/") product {% function(d) { return ["sum",d[0],d[1],d[2]]; } %} | product {% id %}
product -> product ("*"|"/") exp {% function(d) { return ["product",d[0],d[1],d[2]]; } %} | exp {% id %}
exp -> unaryoperation "^" exp {% function(d) { return ["exp",d[0],d[1],d[2]]; } %} | unaryoperation {% id %} # this is right associative!
unaryoperation -> unaryoperation _ ("++"|"--") {% function(d) { return ["unaryoperation",d[0],d[2]]; } %} | mapoperation {% id %}
mapoperation -> mapoperation _ "[" _ expression _ "]" {% function(d) { return ["map",d[0],d[4]]; } %} | element {% id %}
element -> variable {% id %} | number {% id %} | "(" _ expression _ ")" {% function(d) { return d[2]; } %} | "{" _ block "}" {% function(d) { return d[2]; } %} | "if" _ expression _ expression {% function(d) { return ["if",d[2],d[4]]; } %} | "while" _ expression _ expression {% function(d) { return ["for",d[2],d[4]]; } %}
variable -> "$":? [a-z]:+ {% function(d) { return ["variable", d[0], d[1].join("")]; } %}
number -> [0-9]:+ {% function(d) { return ["number", d[0].join("")]; } %}
_ -> ___:* {% id %}
__ -> ___:+ {% id %}
___ -> [\t \n] {% empty %} | mlcomment | slcomment
mlcomment -> "/*" mlcommentchars:+ .:? "*/" {% function(d) { return [];/*["comment", d[1].join("")+d[2]];*/ } %}
slcomment -> "//" [^\n]:* "\n" {% function(d) { return [];/*["comment", d[1].join("")];*/ } %}
mlcommentchars -> "*" [^/] {% function(d) { return d[0] + d[1]; } %} | [^*] . {% function(d) { return d[0] + d[1]; } %}
```

**Comment by Hardmath123** [comment]
This is what we call "syntactic ambiguity": your grammar can be parsed in two different ways that make sense to nearley. For example, `a -> a a | "b"` will parse "bbbbbbb" in many different ways—try it out if this doesn't seem obvious.

Let me know if you get stuck searching for the ambiguity.

**Comment by alecmocatta** [comment]
Thanks. I'm reasonably sure there's no syntactic ambiguity - they're returning identical parse trees.

With this grammar, it returns 4 identical parse trees. Substituting `exp` for `unaryoperation` in the definition of `expression` results in 1 parse – even though `exp -> unaryoperation`.

```
program -> _ block 
block -> (statement _):* 
statement -> expression _ ";" 
expression -> expression _ ("="|"=="|"!="|">"|"<"|"<="|">=") _ exp  | exp 
#expression -> expression _ ("="|"=="|"!="|">"|"<"|"<="|">=") _ unaryoperation  | unaryoperation 
sum -> sum ("*"|"/") product  | product 
product -> product ("*"|"/") exp  | exp 
exp -> unaryoperation
unaryoperation -> unaryoperation _ ("++"|"--")  | mapoperation 
mapoperation -> mapoperation _ "[" _ expression _ "]"  | element 
element -> variable  | number  | "(" _ expression _ ")"  | "{" _ block "}"  | "if" _ expression _ expression  | "while" _ expression _ expression 
variable -> "$":? [a-z]:+ 
number -> [0-9]:+ 
_ -> ___:* 
__ -> ___:+ 
___ -> [\t \n]  | mlcomment | slcomment
mlcomment -> "/*" mlcommentchars:+ .:? "*/" 
slcomment -> "//" [^\n]:* "\n" 
mlcommentchars -> "*" [^/]  | [^*] . 
```

**Comment by Hardmath123** [comment]
Ah, interesting, I see what's going on—it's a glitch I introduced with my previous bugfix. This one's slightly trickier to fix, though. Essentially, the parser is choking on nullables. While I'm busy resolving this, it shouldn't be a hurdle to creating a grammar. Simply ignore the duplicates, they're "harmless".

**Comment by Hardmath123** [comment]
Minimal example:

```
test -> "a" b "\n"
b -> a a
a -> null
```

**Comment by Hardmath123** [comment]
This should be fixed as of 1e8d799812b3ea7b8a49cf77226fa15d73264efb (v1.4.0). Sorry it took a while, I had to read Kegler's paper on Marpa's algorithm. I think I'm doing some approximation of the Aycock-Horsepool algorithm. In any case, all of your test cases work now.

Cheers, and thanks for the bug reports! :-)

**Comment by alecmocatta** [comment]
Thanks for the quick fixes!
